core/src/main/java/hudson/util/io/TarArchiver.java

```diff
@@ -87,23 +91,30 @@ public void visit(File file, String relativePath) throws IOException {
         int mode = IOUtils.mode(file);
         if (mode!=-1)   te.setMode(mode);
         te.setModTime(file.lastModified());
-        if(!file.isDirectory())
-            te.setSize(file.length());
-
-        tar.putArchiveEntry(te);
+        long size = 0;
 
         if (!file.isDirectory()) {
-            FileInputStream in = new FileInputStream(file);
-            try {
-                int len;
-                while((len=in.read(buf))>=0)
-                    tar.write(buf,0,len);
-            } finally {
-                in.close();
+            size = file.length();
+            te.setSize(size);
+        }
+        tar.putArchiveEntry(te);
+        try {
+            if (!file.isDirectory()) {
+                // ensure we don't write more bytes than the declared when we created the entry
+                try (BoundedInputStream in = new BoundedInputStream(new FileInputStream(file), size)) {
+                    int len;
+                    while ((len = in.read(buf)) >= 0) {
+                        tar.write(buf, 0, len);
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    IOException ioE = new IOException("Error writing to tar file from: " + file, e);
+                    ioE.addSuppressed(e);
```

Comment:
:ant: Huh? Why would you call `addSuppressed` on something which is already the `cause`? This makes no sense to me.